File: VIF_git/GSDS-Qual/LeetCode/week2/Graphs_BFS/Rotting_Oranges/suho/RO.cpp
class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        int cnt=0, t=0;
        vector<vector<int>> grid_t, grid_t0;
        grid_t.resize(grid.size());
        grid_t0.resize(grid.size());
        copy(grid.begin(), grid.end(), grid_t.begin());
        copy(grid.begin(), grid.end(), grid_t0.begin());

        while(1){
            int cnt0 = cnt;
            cnt = 0;
            for(int i = 0; i<grid.size(); i++){
                for(int j = 0; j<grid[0].size(); j++){
                    if(grid_t0[i][j]==1){
                        cnt++;
                        if((i-1>=0 && grid_t0[i-1][j]==2) || (j-1>=0 && grid_t0[i][j-1]==2) || (i+1<grid.size() && grid_t0[i+1][j]==2) || (j+1<grid[0].size() && grid_t0[i][j+1]==2)){
                            grid_t[i][j] = 2;
                        }
                    }
                }
            }
            if(cnt==0) return t;
            else if(cnt==cnt0) return -1;
            else{
                copy(grid_t.begin(), grid_t.end(), grid_t0.begin());
                t++;
            }
        }
    }
};

File: VIF_git/GSDS-Qual/LeetCode/week2/Graphs_BFS/Rotting_Oranges/suho/RO.py
import copy
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        t=0
        grid_t = copy.deepcopy(grid)
        grid_t0 = copy.deepcopy(grid_t)
        cnt = 0

        while(1):
            cnt0 = cnt
            cnt = 0
            for i in range(len(grid)):
                for j in range(len(grid)):
                    if(grid_t0[i][j]==1):
                        cnt+=1
                        if( (i-1>=0 and grid_t0[i-1][j]==2) or (j-1>=0 and grid_t0[i][j-1]==2) or (i+1<len(grid) and grid_t0[i+1][j]==2) or (j+1<len(grid[0]) and grid_t0[i][j+1]==2)):
                            grid_t[i][j] = 2

            if cnt == 0 : return t
            elif cnt == cnt0 : return -1
            else : 
                grid_t0 = copy.deepcopy(grid_t)
                t += 1
